### PR TITLE
Move Sphinx-based tutorials to Jupyter notebooks

### PR DESCRIPTION
**Pull request by cdeil**
I'd like to propose that we move the Sphinx-based tutorials we have in `docs/tutorials` here:
http://docs.gammapy.org/en/v0.6/tutorials/index.html
to Jupyter notebooks in `gammapy-extra/notebooks` (and list them at the end under https://nbviewer.jupyter.org/github/gammapy/gammapy-extra/blob/master/index.ipynb#Extra-topics to make it clear that they are pretty specialised, not the first thing to look at for most users, but also not useless IMO).

I'm 50:50 on pros and cons of writing and maintaining tutorials in Sphinx or Jupyter notebooks, so @adonath and @joleroi and anyone - please comment what you think!

The problem with `docs/tutorials` (apart from being mostly old and not super relevant content for most users) is that it's a second place to put tutorials, and it's hard for users or authors to understand what content is where. And I think with the notebooks, people are mostly happy so we don't want to give that up.

There's also the problem that `python setup.py build_docs` is the slowest and most (at least for me) most annoying part of the Gammapy development experience. It currently takes 4 min on my machine, and I'd like to see it get faster, not slower. Currently `docs/tutorials` only takes 20 seconds out of those 4 min, so it's not a big win. But there we were always careful to choose datasets and analyses that run within seconds, whereas in the notebooks it's OK to have an analysis example that takes a bit longer.

We will still have to discuss what content should go in per-subpackage docs like http://docs.gammapy.org/en/latest/spectrum/index.html and what content in separate examples and tutorials. For sure the goal is not to remove all examples from the Sphinx docs and to only leave the API docs there. We will keep both, unless really good tooling comes up that allows generating notebooks from Jupyter notebooks which doesn't exist yet. 

Bottom line - is this aa step in the right direction for now?

### CODE REVIEW

**Comment by adonath** [comment]
I'll think about it a little more, but my first impression is that it is good idea! If I remember correctly most of those tutorials have been added, when `gammapy-extra` hadn't existed yet. I think for users as well as developers it is advantageous to have all the tutorials and detailed usage examples in one place.  Smaller examples are directly included in the API docs anyway, so I think there is few motivation to keep the `docs/tutorials` sub-section. Also from a developers point of view it is much more convenient to run and modify tutorial notebooks, then executing the individual tutorial scripts and copying and pasting code from the docs. 

We could think about replacing the `docs/tutorials` page with an overview on the most relevant tutorials in `gammapy-extra`, that could later transform into a sub-page `gammapy.org/gammapy-tutorials/` similar to http://www.astropy.org/astropy-tutorials/. But I actually think the prominent link we have right now on the starting page http://docs.gammapy.org/ is sufficient...

**Comment by Bultako** [comment]
> We will keep both, unless really good tooling comes up that allows 
> generating notebooks from Jupyter notebooks which doesn't exist yet.

Hi there !
I'm not sure what you mean with **_generating notebooks from Jupyter notebooks_**
In case you're saying _generating **Sphinx docs** from Jupyter notebooks_ I let you know about `nbsphinx` in case you are not aware of it.
http://nbsphinx.readthedocs.io
Cheers

**Comment by cdeil** [comment]
@Bultako - I meant "generating notebooks from RST/Sphinx docs" here. I did look at the existing projects to convert back or forth between notebooks and RST docs pages a while ago, and none seemed like a great solution.

Authoring documentation and tutorials in Jupyter notebooks has the large problem that notebooks are JSON files with ugly diffs, so changes are hard to review in pull requests. So from a content management perspective, storing docs / tutorials in RST files is better. The problem there is that it's cumbersome to write and edit the code there, because one cannot simply execute and try it while editing.

My impression is still that while `nbconvert`, nbdiff` and `nbsphinx` have made progress, it's still painful to maintain docs / tutorials in notebooks, and no standard pattern for maintaining notebooks, and integrating Sphinx docs and notebooks has emerged (that would also allow cross-linking and full-text search on both).

So unless I hear objections here, I'll go ahead with this change (and write / clean up these tutorials as notebooks, one is currently broken, which triggered me to work on this today: https://travis-ci.org/gammapy/gammapy/jobs/283780263#L1923), and then I'm hoping that in the future others will improve the setup / tooling for notebook docs in general, but also for the Gammapy project.

**Comment by Bultako** [comment]
@cdeil 
> from a content management perspective, storing docs / tutorials in RST files is better. The 
> problem there is that it's cumbersome to write and edit the code there, because one cannot
> simply execute and try it while editing.
 
I agree.
I personally use atom+hydrogen to modif and execute lines in *.py scripts _live_ (like if it was a Jupyter notebook) while having the RST file in another atom tab.

I think it's good to separate tutorials as notebooks, and keep small code snippets in docs, as long as docs are complete and do not miss any important info. Maintaining notebooks in a git/diff way is really painful, we cannot avoid that and I would not do it, but I like very much to have them. Yeah, I'm certain content management in notebooks-format will improve at some moment with versioning, search features, doc-formatting, etc..

**Comment by adonath** [comment]
Here are a few thoughts on the question which content should be part of the per-subpackage docs and which content should be in separate examples and tutorials:

* I think examples in the API docs should be short, self-contained, minimal examples, that typically illustrate a single technical aspect and the usage of the API with a few lines of code. No scientific example data is needed or relevant for the understanding. It typically shows the usage of **one class or method**. Ideally those examples do not require large datasets from `gammapy-extra` and can be examples with "toy data" as well.

* Tutorials should have a scientific context and/or application, which goes beyond the technical aspect of teaching the API (even though it is also a goal...). They typically combine **several classes and methods** to illustrate how the different parts of gammapy work together and higher level analysis chains can be build. It uses real or simulated scientific data.

* Still there are a few corner cases e.g. a tutorial that focuses purely on a single class or method and explains many aspects of it in detail e.g. a tutorial on how to work with sky images in general. There I don't really have an opinion, I think both `gammapy-extra/notebooks` as well as `docs/image` would be a good place.

**Comment by cdeil** [comment]
@adonath - Just one addition to what you said: I think it's important to have extensive x-linking between different levels of docs, and also be able to search them. One thing I like about the scikit-learn docs is that all their class-level docstrings end with "Read more in the User Guide.", i.e. with a link to high-level docs. Example: http://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html

So we need some tooling to go between IPYNB and RST. Probably using `nbconvert` / `sphinxnb` (the Astropy tutorials are switching to `sphinxnb` at the moment).

@joleroi - You OK with this change for now?

**Comment by joleroi** [comment]
> @joleroi - You OK with this change for now?

:+1:

**Comment by cdeil** [comment]
Merging now. I'll add that stuff back as notebooks later this week.

**Comment by Bultako** [comment]
@cdeil 

> So we need some tooling to go between IPYNB and RST. Probably using 
> nbconvert / sphinxnb (the Astropy tutorials are switching to sphinxnb at
> the moment).

I propose here an initial idea to face it:

**Linking RST Docs -> RST Notebooks**

_RST notebooks_ could be created with `nbsphinx` and moved to a `notebooks` subfolder in docs. 

This provides the possibility to link from _RST Docs_ -> _RST notebooks_  and make  _RST notebooks_ searchable. 

**Linking RST Notebooks -> IPYNB Notebooks**

We could develop a small script to launch in the last step of the "RST notebooks creation process", in order to add systematically a footer linking to the correspondent _Raw IPYNB notebook_ in GitHub `gammapy-extra` repo (or maybe should we also include the *.ipynb files in the gammapy distribution?)

**Linking IPYNB Notebooks -> HTML Docs and HTML Notebooks** (closing the circle)

The same small script above could also add footer/header cells to the ipynb files with links to a particular or common URL in http://docs.gammapy.org.

----

* As long as *.ipynb notebooks remain outside the gammapy distribution, I'm afraid an internet connection is needed to link *.ipynb files in `gammapy-extra` repo with the published doc.

* There will be new (automated) actions to make in the management of notebooks, simply merging in GitHub will not suffice:

          - creation of *.rst notebooks files
          - modification of *.ipynb notebooks files

**Comment by cdeil** [comment]
@Bultako - I agree, something very similar to what you propose is the way to go for Gammapy docs (and maybe be a shining example for other projects on how to have a half RST, half IPYNB docs system). 

I don't think a single-repo approach is a good idea. We already have `gammapy-extra` with the notebooks and example data files, because we wanted to keep the main `gammapy` code repo small. Similarly, PyPI has a limit of ~ 50 MB for distributing Python packages, and I don't think we should try to distribute notebooks that way. For documentation build, a more complex setup is needed. We could try to build everything on `readthedocs`, or if that doesn't work well (we already had many install and timeout issues there), then we could move the Gammapy docs build to a Jenkins we have running here at MPIK. 

Note that there are already some Python scripts to test notebooks:
https://github.com/gammapy/gammapy/blob/master/test_notebooks.py
and there is a little bit of tooling already to link from Sphinx docs to notebooks (but not to tags or sections yet):
https://github.com/gammapy/gammapy/blob/master/docs/conf.py#L162
https://github.com/gammapy/gammapy/blob/master/docs/conf.py#L178

The main problem to move forward is that someone needs to spend the time to implement something better. I think even for someone knowledgeable with Sphinx and nbconvert, it might be a few  days of work to get everything working. (See how at https://github.com/astropy/astropy-tutorials/pull/153 Adrian is doing it for Astropy, and it takes time because there's missing features and issues in `sphinxnb` and `nbconvert` and it takes time to figure out the details).

I am tempted, but won't make this my project in the near future. @Bultako or anyone interested in notebook and docs tooling - if you have some time and want to give it a shot -- I'm available for a Skype call or further discussion / reviewing pull requests in that direction.

**Comment by Bultako** [comment]
I guess I could give it a try. 
I would need some info on the details of how you proceed with the process of the doc building. @cdeil I have sent to you a Skype friendship request if you and others want to chat/talk about this.

**Comment by Bultako** [comment]
FYI

I have done a very quick test to build `gammapy` notebooks in sphinx format with `nbsphinx` and `astropy` sphinx theme. I have not found any big issues, just several notebooks with errors (which cause the process to stop) You may find the results in:

http://www.iaa.es/~jer/gammapydocs/home.html

**Comment by cdeil** [comment]
@Bultako - This looks great to me, thanks!

I think we should try to add this to http://docs.gammapy.org/ somehow.

The main advantages that we gain that I see now that we gain for notebooks are:
- Search
- Table of content / navigation via sidebar

And there's a few advantages that I think we'll get, even though I don't know exactly how it works yet:
- Easier cross-linking between notebooks and Sphinx docs (see nbsphinx docs)
- We introduce a "build" or "render" steps for the notebooks, so in the future we could store notebooks without output in the repo, or have them all rendered and styled as we like (no longer depends on matplotlib version and settings on developer machine that last executed the notebook)

---

@Bultako - Some points for discussion:

- Can you please share the script or what you did, so that I can try it out as well? Maybe in a separate Github repo for now? Or directly as a pull request against the Gammapy repo?

- How would the notebooks relate to our existing Sphinx docs, that are at http://docs.gammapy.org/ and can be built locally via `python setup.py build_docs`? Would it in the end be one Sphinx project or two separate ones?

- How long does it take to do the sphinx build for the notebooks? I looked at the `nbsphinx` docs, and see that one can configure if Notebooks are executed or not. Maybe for now, the easiest would be to try and run this on readthedocs and have it be part of the one Sphinx project at http://docs.gammapy.org/ ? I think if we do it this way, then we should not re-run any notebooks (would take long and time out and require all datasets to be on readthedocs), but just use `nbsphinx` to convert them. How long does this take roughly? A minute or much more?

- The alternative is that we try to set up a Jenkins job or some other CI to execute and render all the notebooks from time to time, and figure out a way to host them (I don't think readthedocs allows upload, so probably via github-pages). Seems like a lot of work and we get most of the gains with the simpler option mentioned in the last point, no?

---

PS: a nice example I found:
http://yt-project.org/doc/quickstart/index.html
I think they have just Python scripts using `nbconvert` directly, not going via `nbsphinx` though.
From what I gather, for us, `nbsphinx` would be a good way to go.

**Comment by Bultako** [comment]
@cdeil 

- I have not done any scripting for the moment. Just launching Sphinx in the notebooks folder after slight modify in Sphinx conf.py to use `nbsphinx`and astropy theme.
I will move the whole to a separate GitHub repo.

- I think the best is to build the _HTML notebooks_ in the same `python setup.py build_docs` doc generation process. I think it should not be very hard, but I would need some small intro/guidance in the peculiarities of `astropy` helpers used and config. So my goal is to have a single Sphinx project. Here the readme: http://www.iaa.es/~jer/gammapydocs/README.html

- I think most of the notebooks that I have used contain pre-executed output cells. The default behavior of `nbsphinx`is to execute only cells with no output. In my quick test the whole conversion process did not take more that 30 sec. I got several notebooks stopping the process because of output errors in cells, which mean that some of them where actually executing live in the conversion process. Yes, we could try `nbsphinx`in `readthedocs` with pre-executed notebooks. This actually solves many other issues. http://nbsphinx.readthedocs.io/en/0.2.14/pre-executed.html 

I'll be back to you whith a GitHub repo for this

**Comment by cdeil** [comment]
@Bultako - Agree 100%. Thanks!

I'm not very familiar with astropy-helpers and `python setup.py build_docs`, I would also have to start by reading how it's implemented and look at the existing examples how to extend it.

It would be nice if one could execute the notebook part of the sphinx build separately. If later we change that part to also execute notebooks, then it's even required, so that someone working on the docstrings in the Gammapy code doesn't need to wait for that (ideally imports are even delayed so that the normal sphinx build keeps working without having nbsphinx and pandoc installed).

@Bultako - I think I already mentioned it, but just in case: note that we already have some sphinx / notebook-related scripts in  https://github.com/gammapy/gammapy/blob/master/gammapy/utils/docs.py and https://github.com/gammapy/gammapy/blob/master/test_notebooks.py . If there is more notebook / nbsphinx related code / tooling we need to write, I think it should go in `gammapy.utils` if possible, but other new files at the top level or in the docs folder are OK as well.

**Comment by cdeil** [comment]
PS: I'll go through our existing notebooks now and do some formatting improvements based on what I see at http://www.iaa.es/~jer/gammapydocs/home.html

E.g. for the "first steps with Gammapy" the table of content wasn't good because headings weren't at the right level. This is fixed in  https://github.com/gammapy/gammapy-extra/commit/cd9cef9373836ef55923d0e26247073222aabf53

So the Sphinx tutorial notebooks should come out even more pretty / useful. :-)

**Comment by Bultako** [comment]
I have forked a gammapy repo in my GitHub profile, and created a branch `docbuilding-notebooks` https://github.com/Bultako/gammapy/tree/docbuilding-notebooks

Main modifs are: 
* .gitignore to track 
       - the generated doc just for your inspection in `docs/_build`folder
       - the *.ipynb files in `docs/folder` (these will be copied from GAMMAPY-EXTRA later on)
* conf.py Sphinx config file in `docs` to make `nbconfig` work nicely
* some *.rst files in `docs`and some *.py files in `gammany`source code folder to replace directive `.. code::` by `..code :: python`and make code inclusion in docs working. You also need to re-install gammany with these new source files.

Remember you need to install `nbsphinx`:
conda install -c conda-forge nbsphinx

And that's all for the moment.
Just type `python setup.py build-docs` and this mixed interlinked doc of *rst files and notebooks is generated in one shot. We still have some warnings die to orphans *.ipynb files, but this does not prevent their conversion.

This is the firs step, still work in progress..

**Comment by cdeil** [comment]
@Bultako - I'll have a look soon.

My suggestion would be that you open a PR against the Gammapy repo soon.
It's better to have discussion record in the PR, instead of continuing here in this closed other PR.
Also, if it's OK with you, I would like to aim to have a first version merged very soon, so that we can see if we can get it to run on readthedocs or not. It doesn't matter if some thing don't work yet, we can go step by step.

**Comment by Bultako** [comment]
Done
I have removed the html docs generated in `docs/_build' 

https://github.com/gammapy/gammapy/pull/1176
